File: apps/api/app/routers/insight_tasks.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.integrations.supabase import require_supabase

router = APIRouter(prefix="/api/v1/insight-tasks", tags=["insight-tasks"])

ALLOWED_STATUS = {"pending", "running", "success", "failed", "cancelled"}


class InsightTaskCreate(BaseModel):
    platform: str = Field(min_length=1, max_length=64)
    product_id: str = Field(min_length=1, max_length=256)
    analysis_provider_id: str | None = Field(default=None, max_length=128)


@router.get("")
def list_insight_tasks() -> dict:
    try:
        sb = require_supabase()
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e)) from e
    try:
        res = (
            sb.table("insight_tasks")
            .select("*")
            .order("created_at", desc=True)
            .limit(200)
            .execute()
        )
    except Exception as e:  # noqa: BLE001
        raise HTTPException(
            status_code=502,
            detail=f"Supabase 查询失败：{e!s}",
        ) from e
    return {"items": res.data or []}


@router.post("")
def create_insight_task(body: InsightTaskCreate) -> dict:
    try:
        sb = require_supabase()
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e)) from e
    row = {
        "platform": body.platform.strip(),
        "product_id": body.product_id.strip(),
        "status": "pending",
        "analysis_provider_id": body.analysis_provider_id,
    }
    try:
        res = sb.table("insight_tasks").insert(row).select("*").execute()
    except Exception as e:  # noqa: BLE001
        raise HTTPException(
            status_code=502,
            detail=f"Supabase 写入失败：{e!s}",
        ) from e
    data = res.data
    if not data:
        raise HTTPException(status_code=500, detail="插入成功但未返回数据，请检查 RLS/策略")
    return data[0]

File: packages/contracts/src/insight-tasks.ts
/** 与 Supabase `insight_tasks` 表及 GET /api/v1/insight-tasks 对齐 */

export type InsightTaskStatus =
  | 'pending'
  | 'running'
  | 'success'
  | 'failed'
  | 'cancelled'

/** Pangolin amzProductDetail 等写入 insight_tasks.product_snapshot */
export type InsightTaskProductSnapshot = {
  title?: string | null
  image_url?: string | null
  price_display?: string | null
  asin?: string | null
  source?: string | null
  fetched_at?: string | null
}

export type InsightTaskRow = {
  id: string
  platform: string
  product_id: string
  status: InsightTaskStatus
  analysis_provider_id: string | null
  /** 词典类目：general / electronics 等 */
  dictionary_vertical_id?: string | null
  /** 创建人用户名 */
  created_by?: string | null
  error_code: string | null
  error_message: string | null
  failure_stage: string | null
  created_at: string
  updated_at: string
  product_snapshot?: InsightTaskProductSnapshot | null
}

export type InsightTaskListResponse = {
  items: InsightTaskRow[]
}

export type InsightTaskCreateBody = {
  platform: string
  product_id: string
  analysis_provider_id?: string | null
  /** 默认 general；与后端词典垂直一致 */
  dictionary_vertical_id?: string
}

/** PATCH /api/v1/insight-tasks/:id — 状态迁移需符合服务端状态机 */
export type InsightTaskPatchBody = {
  status: InsightTaskStatus
  error_code?: string | null
  error_message?: string | null
  failure_stage?: string | null
}
